resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/builders/AutoBuildJobRescheduleTest.java

```diff
@@ -0,0 +1,200 @@
+		public boolean contains(ISchedulingRule rule) {
+			return rule == this || rule == file;
+		}
+	}
+
+	/**
+	 * An {@link IJobChangeListener} that reacts to the AutoBuildJob being scheduled
+	 * or done.
+	 *
+	 * Since the AutoBuildJob is not visible, we check against the class name.
+	 */
+	private IJobChangeListener jobChangeListener = new JobChangeAdapter() {
+		@Override
+		public void scheduled(IJobChangeEvent event) {
+			if ("AutoBuildJob".equals(event.getJob().getClass().getSimpleName())) {
+				rescheduled.set(true);
+			}
+		}
+
+		@Override
+		public void done(IJobChangeEvent event) {
+			if ("AutoBuildJob".equals(event.getJob().getClass().getSimpleName())) {
+				barrier.setStatus(AUTOBUILD_DONE);
+			}
+		}
+	};
+
+	public AutoBuildJobRescheduleTest(String testName) {
+		super(testName);
+	}
+
+	/**
+	 * Setup the test by creating a project and a file. We also add a special
+	 * builder that suspends execution using the {@link #barrier}.
+	 */
+	@Override
+	protected void setUp() throws Exception {
+		super.setUp();
+
+		var project = getWorkspace().getRoot().getProject("PROJECT");
+		file = project.getFile("file");
+		try {
+			// Turn auto-building off to prevent too early builds
+			setAutoBuilding(false);
+			// Create and open the project
+			project.create(getMonitor());
+			project.open(getMonitor());
+			// Create the file
+			ensureExistsInWorkspace(file, getRandomContents());
+
+			// Add the EmptyDeltaBuilder. The behavior of that builder is configured below.
+			IProjectDescription desc = project.getDescription();
+			desc.setBuildSpec(new ICommand[] { createCommand(desc, EmptyDeltaBuilder.BUILDER_NAME, "ProjectBuild") });
+			project.setDescription(desc, getMonitor());
+			project.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, getMonitor());
+
+			// Perform the first build. This instantiates the EmptyDeltaBuilder and makes
+			// its instance accessible.
+			setAutoBuilding(true);
+
+			// Configure the behavior of the builder
+			EmptyDeltaBuilder.getInstance().setRuleCallback(new BuilderRuleCallback() {
+				// We only execute the block-continue on the first run in the test case.
+				// Therefore, we remember whether we have already run with this flag.
+				private AtomicBoolean buildRan = new AtomicBoolean(false);
+
+				@Override
+				public IProject[] build(int kind, Map<String, String> args, IProgressMonitor monitor)
+						throws CoreException {
+					// only on the first run, block the builder
+					if (!buildRan.getAndSet(true)) {
+						barrier.setStatus(AUTOBUILD_RUNNING_BLOCKED);
+						// block until the WorkspaceJob has made the file dirty
+						barrier.waitForStatus(AUTOBUILD_CONTINUE);
+					}
+					return new IProject[] { project };
+				}
+			});
+
+			// Wait until the Content Description Update job has finished, because it also
+			// might schedule a build and we don't want that interfere with our test
+			// process.
+			waitForContentDescriptionUpdate();
+		} catch (CoreException e) {
```

Comment:
setUp() throws Exception
why convert the exception?

Reply:
the other tests sometimes did that, so this got in when I based my code on the other tests. But you are correct, it makes no sense here.